api/chat.go

```diff
@@ -308,70 +318,75 @@ func (r chatRoutes) handleApprove(ctx tools.TUMLiveContext, msg []byte) {
 	broadcastStream(ctx.Stream.ID, broadcastBytes)
 }
 
-func (r chatRoutes) handleRetract(ctx tools.TUMLiveContext, msg []byte) {
-	var req wsIdReq
+func (r chatRoutes) handleReactTo(ctx tools.TUMLiveContext, msg []byte) {
+	var req wsReactToReq
 	err := json.Unmarshal(msg, &req)
 	if err != nil {
-		log.WithError(err).Warn("could not unmarshal message retract request")
-		return
-	}
-	if ctx.User == nil || !ctx.User.IsAdminOfCourse(*ctx.Course) {
+		log.WithError(err).Warn("could not unmarshal reactTo request")
 		return
 	}
 
-	err = r.ChatDao.RetractChat(req.Id)
-	if err != nil {
-		log.WithError(err).Error("could not retract chat")
+	if _, isAllowed := allowedReactions[req.Emoji]; !isAllowed {
+		log.Warn("user tried to add illegal reaction")
 		return
 	}
 
-	err = r.ChatDao.RemoveLikes(req.Id)
+	err = r.ChatDao.ToggleReaction(ctx.User.ID, req.wsIdReq.Id, ctx.User.Name, req.Emoji)
 	if err != nil {
-		log.WithError(err).Error("could not remove likes from chat")
+		log.WithError(err).Error("error reacting to message")
 		return
 	}
-
-	chat, err := r.ChatDao.GetChat(req.Id, 0)
+	reactions, err := r.ChatDao.GetReactions(req.Id)
 	if err != nil {
-		log.WithError(err).Error("could not get chat")
+		log.WithError(err).Error("error getting num of chat reactions")
+		return
 	}
 	broadcast := gin.H{
-		"retract": req.Id,
-		"chat":    chat,
+		"reactions": req.Id,
+		"payload":   reactions,
 	}
 	broadcastBytes, err := json.Marshal(broadcast)
 	if err != nil {
-		log.WithError(err).Error("could not marshal retract message")
+		log.WithError(err).Error("Can't marshal reactions message")
 		return
 	}
 	broadcastStream(ctx.Stream.ID, broadcastBytes)
 }
 
-func (r chatRoutes) handleLike(ctx tools.TUMLiveContext, msg []byte) {
+func (r chatRoutes) handleRetract(ctx tools.TUMLiveContext, msg []byte) {
 	var req wsIdReq
 	err := json.Unmarshal(msg, &req)
 	if err != nil {
-		log.WithError(err).Warn("could not unmarshal like request")
+		log.WithError(err).Warn("could not unmarshal message retract request")
+		return
+	}
+	if ctx.User == nil || !ctx.User.IsAdminOfCourse(*ctx.Course) {
 		return
 	}
 
-	err = r.ChatDao.ToggleLike(ctx.User.ID, req.Id)
+	err = r.ChatDao.RetractChat(req.Id)
 	if err != nil {
-		log.WithError(err).Error("error liking/unliking message")
+		log.WithError(err).Error("could not retract chat")
 		return
 	}
-	numLikes, err := r.ChatDao.GetNumLikes(req.Id)
+
+	err = r.ChatDao.RemoveReactions(req.Id)
```

Comment:
Do we need to remove reactions when a chat is retracted?

Reply:
We remove the likes of a message if it is retracted. So yes, I think to stay consistent we also need to delete reactions on retraction. 👍🏻